### ISSUE TITLE
Resolve master merge contention caused by up-to-date and CI branch protections

### ISSUE THREAD

**Issue by anorth** [filecoin-project/go-filecoin#2057]
### Description
We use GitHub branch protections that require a PR to be merged to master to:
- be up to date with master (the commits in the branch are directly based on master HEAD)
- have a passing CI

These combine to mean that CI must pass at the exact commit to be merged. CI currently takes about 12 minutes. This means:
- PRs must be processed in strict serial order, with a painful delay minutes between them
- Handling minor review comments requires pushing and waiting for CI again
- If anyone else commits, all outstanding PRs must rebase and re-run CI

Our async workflows mean it's quite common to have multiple in-flight PRs. Wrangling a collection of approved PRs to master now takes a long time. It's easy to lose attention and be pushed back by other commits landing, and the 12-minute blocks of time in between merges are not very useful for other work.

I consider this a huge drain on productivity and, at 10+ minute CI, approaching "on fire" in urgency. (I recently spent about two hours landing six PRs, only one of which involved nontrivial rebase work).

Some options for resolving this include:
- Lift the up-to-date branch protection, relying on committers' judgement when merging PRs with passing build prior to rebase/fixups. This opens the possibility of build breakages going to master. It could be accompanied by a no-questions-asked revert policy when that happens
- Lift the CI branch protection, relying on committers' judgement and local testing (the CI integration would remain, but be informational rather than required). This opens the possibility of build breakages on master.
- Make CI *much* faster, so that an update can be tested in <= ~2 minutes
- Build/use a merge queue tool to automate the landing of PRs by sequentially rebasing, testing, and merging.

Committers' judgement could be enhanced through local tooling such as commit/push hooks.

See #1976 for some CI speed-up, possibly sufficient to reduce this from flaming to smouldering for now.

### Acceptance criteria
- Developers reliably spend <= 2 minutes landing an approved PR to master, unless automated rebasing or tests fail and require manual intervention
- Broken builds on master are sufficiently rare and short-lived (for some pragmatic definition of sufficient)

### Risks + pitfalls
- Relaxing constraints risks broken builds sapping the productivity of other developers. In rare situations, rolling back may not immediately unblock people.
- A merge queue is extra effort and operational overhead
- Making CI faster now is a short-lived fix; it will get slower again

### Where to begin
Discuss.

**Comment by anorth** [comment]
A merge queue can achieve total confidence in green builds on master by automating around strict checks. If you haven't used one before, articles [here](https://medium.com/strava-engineering/butler-merge-queue-how-strava-merges-code-7095a3310930), [here](https://docs.microsoft.com/en-us/azure/devops/learn/git/race-to-push) and [here](https://bors.tech/devdocs/bors-ng/readme.html) describe something like what I'm thinking of, though I don't think any solve it for us immediately.

It *may* be possible to hack something like this into Circle+GitHub without adding additional infrastructure, though a good solution needs some idea of a canonical queue state. Circle would need push access for this.

It's probably a bunch of work (2-3 eng weeks?) to implement something robust here. Depending on team size and hours lost per week, this could pay off in a few months. A great solution might look like an open source project, separate from Filecoin, that others can use and contribute to too.

**Comment by rosalinekarr** [comment]
@ognots @gmasgras, opinions?

**Comment by ognots** [comment]
## On CircleCI build time

We could make significant gains in build time, but <= 2m is very difficult presently.

Here's a list of the top 10 unit test durations taken from `Test` step

```sh
$ cat ~/Downloads/build_11317_step_119_container_0.txt | grep 's)' | cut -d':' -f2 | cut -d' ' -f3 | cut -d '(' -f2 | cut -d ')' -f1 | sort -gr | head -n10
332.17s
107.05s
89.02s
53.01s
50.08s
33.01s
10.48s
8.22s
7.68s
7.47s
```
Even with a high degree of parallelism, these numbers alone make the < 2m goal challenging

Some other longer duration jobs in CircleCI build:
*  `Install Dependencies` - 1m04s
* `Restoring cache` - ~30s each x 2

### Possible efficiencies

* investigating how CircleCI parallelism might affect these numbers
* splitting `Test`, `Functional Test` and `Lint` steps in to separate jobs with workflow configured to run these jobs concurrently
* look for ways to reduce duration of `Install Dependencies`

### Conclusion

My feelpinion on the matter is that it's definitely worth investigating ways to reduce build time to help mitigate merge contention

**Comment by ognots** [comment]
## On up-to-date branch protection

The risk factor is low enough to seriously consider this path. 

In the [test_all workflow](https://github.com/filecoin-project/go-filecoin/blob/master/.circleci/config.yml#L403-L412) a failure of `build_linux` on merging to `master` will prevent any artifact from being published, which presently is a docker image.

While a broken master can cause hiccups for anyone rebasing off of it while in a broken state, a revert is a fast enough solution.

Pull Requests report if a feature branch is able to be merged in to master without conflict which is at least some level of protection.

The impacted stakeholder are developers only

A drawback I've seen in the past with this pattern is the need for developers to track their specific master build status after merging

**Comment by ognots** [comment]
## On Merge Queue

No objections in principle.

I would agree on the assessment of 2-3 dev weeks, but it's also important to consider the additional operational overhead and adding a point of failure to the dev work flow.

My feelpinion on this would be to give it a lower priority at this time

**Comment by dignifiedquire** [comment]
I woul recommend looking into a bot that auto merges on ci passing, and possibly tries to auto rebase if possible.

Ther rust project uses a bot that is opensource afaik so you could take some inspiration there. https://github.com/graydon/bors

**Comment by phritz** [comment]
in the short term we should probably lift the up to date branch protection requirement and accept the risk. we should also have a better solution, like a submit queue. this is a great project for someone with clear impact, i will bring it to infra prioritization next week. open question whether we should try to optimize test time down, my sense is no unless scoped really tightly like spend 3 points on it.

i would like to ensure that this issue here 2057 is holistic in in scope. i would like to clearly establish goals for git history as part of it, encode mechanisms for enforcing / facilitating those goals in gh, and ensuring the contributing guide makes the expectations really clear. the things in this are that i think are unclear are as follows:
- is our goal to avoid merge commits in master, i think the answer is yes, or at least my answer is yes. lets enshrine this in policy.
- are merge commits created when someone uses the 'update branch' button in gh still present when you squash and merge?
- are people not squashing their commits when merging to master, and is this contributing to the problem?
- are there defaults or preferences that we should set in the project or personal gh settings that help people do the right thing by default? if so we should set them and they should go in the contributing guide. 
- is there some rebase merge option that i should have been using but haven't?

**Comment by phritz** [comment]
assigning to anorth so we know who is going to follow up.

**Comment by phritz** [comment]
i have removed the 'Require branches to be up to date before merging' requirement from master for the time being

**Comment by phritz** [comment]
note https://github.com/filecoin-project/go-filecoin/issues/1976 is also in progress

**Comment by laser** [comment]
On slow go-filecoin tests:

Those SectorBuilder tests (`interface_test.go`) perform (at least one) real seal, unseal, proof-of-spacetime generation and verification cycle through CGO. So, each one of those tests will always take upwards of a minute. 

Some ideas for how we could reduce the impact that the `interface_test.go` have on CircleCI build times:

1. Delete those tests and instead rely on the very high-level functional tests (risky!)
1. Modify CircleCI config such that it runs those tests on a schedule (daily, hourly, etc.) instead of on every push

**Comment by phritz** [comment]
@laser re 2 isn't that what #1976 entails, which is already in progress?

**Comment by laser** [comment]
> @laser re 2 isn't that what #1976 entails, which is already in progress?

Yep, that's right. The `#2` item is what #1976 entails, and #1976 is already in progress.

**Comment by anorth** [comment]
We can configure this repo to [disallow merge commits](https://help.github.com/en/articles/configuring-commit-squashing-for-pull-requests) (at least from being created via GitHub PR buttons), requiring a rebase or squash (which also rebases).

I propose we trial this for a week, then decide whether to keep it. @mishmosh could you please make that change and announce the trial in #fil-dev?

**Comment by mishmosh** [comment]
Merge commits are now disallowed, feedback on impacts is encouraged. We'll re-evaluate after a week.

**Comment by anorth** [comment]
Summary:
- The "up to date" branch protection has been lifted, so concurrent merges don't compete for CI time
- CI is down to about six minutes, though I hope we can keep pushing on this
- A policy discouraging merge commits has been documented in CONTRIBUTING.md and experimentally enforced on GitHub

It would be ideal to either have a merge queue and/or CI that runs tests in only a couple of minutes, but the contention has been lifted and so I'm closing this issue.
